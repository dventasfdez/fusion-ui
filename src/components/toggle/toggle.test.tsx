import React from 'react';
import Toggle from './toggle';
import renderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';
import {fireEvent, render} from '@testing-library/react';
import '@testing-library/jest-dom';

const ToggleCombinations = () => {
  return (
    <>
      <Toggle className="tag-ds" />
      <Toggle className="tag-ds" id="small-toggle" small />
      <Toggle className="tag-ds" id="toggle-checked" checked />
      <Toggle className="tag-ds" id="small-toggle-checked" small checked />
      <Toggle className="tag-ds" id="toggle-disabled" disabled />
      <Toggle className="tag-ds" id="small-toggle-disabled" small disabled />
      <Toggle className="tag-ds" id="toggle-readOnly" readOnly />
      <Toggle className="tag-ds" id="small-toggle-readOnly" small readOnly />
      <Toggle className="tag-ds" id="toggle-checked-disabled" checked disabled />
      <Toggle className="tag-ds" id="small-toggle-checked-disabled" small checked disabled />
      <Toggle className="tag-ds" id="toggle-label" label="I am label for #toggle-label" />
      <Toggle className="tag-ds" id="small-toggle-label" small label="I am label for #toggle-label" />
      <Toggle className="tag-ds" id="toggle-helper-text" helperTextOff="off" helperTextOn="on" />
      <Toggle className="tag-ds" id="small-toggle-helper-text" small helperTextOff="off" helperTextOn="on" />
    </>
  );
};

test('Toggle combinations render', () => {
  const component = renderer.create(<ToggleCombinations />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Click action toggle component', () => {
  const {container, getByTestId} = render(<Toggle data-testid="toggle-test" helperTextOn="on" helperTextOff="off" />);
  const toggleContainer = container.querySelector('.toggle')?.querySelector('.toggle-container');
  if (toggleContainer) {
    expect(toggleContainer.querySelectorAll('.toggle-pill')?.length).toBe(1);
    const togglePill = toggleContainer.querySelector('.toggle-pill');

    if (togglePill) {
      const toggle = getByTestId('toggle-test-input');
      if (toggle) {
        expect(toggle).not.toBeChecked();
        ReactTestUtils.Simulate.change(toggle);
        expect(toggleContainer.querySelectorAll('.toggle-pill_checked').length).toBe(1);
        expect(toggleContainer.querySelector('.toggle-pill_checked')?.querySelectorAll('#toggle').length).toBe(1);
        expect(toggleContainer.querySelector('.toggle-pill_checked')?.querySelector('#toggle')).toBeChecked();

        ReactTestUtils.Simulate.change(toggle);

        expect(toggleContainer.querySelectorAll('.toggle-pill')?.length).toBe(1);
        expect(toggle).not.toBeChecked();
      }
    }
  }
  document.body.removeChild(container);
});

test('Click on toggle with onChange (non-function)', () => {
  const onChange = 'string';
  // @ts-expect-error non-function onChange
  const {getByTestId} = render(<Toggle data-testid="toggle-test" onChange={onChange} />);
  // const toggleContainer = container.querySelector('.toggle')?.querySelector('.toggle-container');
  fireEvent.click(getByTestId('toggle-test'));
  expect(onChange).toBe('string');
});

test('Click on toggle with onChange (function)', () => {
  const onChange = jest.fn();
  const {getByTestId} = render(<Toggle data-testid="toggle-test" onChange={onChange} />);
  fireEvent.click(getByTestId('toggle-test-input'));
  expect(onChange).toBeCalled();
});

test('Toggle change checked prop', () => {
  const {rerender, getByTestId} = render(<Toggle data-testid="toggle-test" />);
  expect(getByTestId('toggle-test')).not.toBeChecked();
  rerender(<Toggle data-testid="toggle-test" checked />);
  expect(getByTestId('toggle-test-input')).toBeChecked();
});

test('Toggle with no testId', () => {
  const _ = render(<Toggle readOnly disabled />);
  // test unable to find an elment by testId
  expect(_).toBeTruthy();
  // expect(() => getByTestId('toggle-test')).toThrow();
});
