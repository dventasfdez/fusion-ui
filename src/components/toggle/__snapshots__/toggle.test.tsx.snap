// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle combinations render 1`] = `
Array [
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill"
        htmlFor="toggle"
        id="toggle-label"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="toggle-input"
          id="toggle"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill_small"
        htmlFor="small-toggle"
        id="small-toggle-label"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="toggle-input"
          id="small-toggle"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill_checked"
        htmlFor="toggle-checked"
        id="toggle-checked-label"
        onClick={[Function]}
      >
        <input
          checked={true}
          className="toggle-input"
          id="toggle-checked"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill_small_checked"
        htmlFor="small-toggle-checked"
        id="small-toggle-checked-label"
        onClick={[Function]}
      >
        <input
          checked={true}
          className="toggle-input"
          id="small-toggle-checked"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill_disabled"
        htmlFor="toggle-disabled"
        id="toggle-disabled-label"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="toggle-input"
          disabled={true}
          id="toggle-disabled"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill_small_disabled"
        htmlFor="small-toggle-disabled"
        id="small-toggle-disabled-label"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="toggle-input"
          disabled={true}
          id="small-toggle-disabled"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill_disabled"
        htmlFor="toggle-readOnly"
        id="toggle-readOnly-label"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="toggle-input"
          id="toggle-readOnly"
          onChange={[Function]}
          readOnly={true}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill_small_disabled"
        htmlFor="small-toggle-readOnly"
        id="small-toggle-readOnly-label"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="toggle-input"
          id="small-toggle-readOnly"
          onChange={[Function]}
          readOnly={true}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill_checked_disabled"
        htmlFor="toggle-checked-disabled"
        id="toggle-checked-disabled-label"
        onClick={[Function]}
      >
        <input
          checked={true}
          className="toggle-input"
          disabled={true}
          id="toggle-checked-disabled"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill_small_checked_disabled"
        htmlFor="small-toggle-checked-disabled"
        id="small-toggle-checked-disabled-label"
        onClick={[Function]}
      >
        <input
          checked={true}
          className="toggle-input"
          disabled={true}
          id="small-toggle-checked-disabled"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <p
      className="toggle-label"
    >
      I am label for #toggle-label
    </p>
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill"
        htmlFor="toggle-label"
        id="toggle-label-label"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="toggle-input"
          id="toggle-label"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill_small"
        htmlFor="small-toggle-label"
        id="small-toggle-label-label"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="toggle-input"
          id="small-toggle-label"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill"
        htmlFor="toggle-helper-text"
        id="toggle-helper-text-label"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="toggle-input"
          id="toggle-helper-text"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
      <span
        className="toggle-text-helper"
      >
        off
      </span>
    </div>
  </div>,
  <div
    className="toggle tag-ds"
  >
    <div
      className="toggle-container"
    >
      <label
        aria-label="toggle-label"
        className="toggle-pill_small"
        htmlFor="small-toggle-helper-text"
        id="small-toggle-helper-text-label"
        onClick={[Function]}
      >
        <input
          checked={false}
          className="toggle-input"
          id="small-toggle-helper-text"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="toggle-handle"
        />
      </label>
      <span
        className="toggle-text-helper_small"
      >
        off
      </span>
    </div>
  </div>,
]
`;
